tachyon/zk/plonk/halo2/optimizations.md

```diff
@@ -0,0 +1,91 @@
+[permuted table poly]: https://github.com/scroll-tech/halo2/blob/1070391642dd64b2d68b47ec246cba9e35bd3c15/halo2_proofs/src/plonk/lookup/prover.rs#L139-L141
+[lookup product poly]: https://github.com/scroll-tech/halo2/blob/1070391642dd64b2d68b47ec246cba9e35bd3c15/halo2_proofs/src/plonk/lookup/prover.rs#L291-L293
+
+## 2. Maximizing Batch Normalize
+
+Originally, when committing to multiple polynomials sequentially, the conversion from bucket type to affine point is done individually for each commitment, which leads to a large number of **expensive** inverse operations. We optimize this inefficiency with **Batch Normalize**.
+
+To set up **Batch Normalize**, conversions from bucket type to affine point are delayed until all commitments are calculated. Once all calculations are finished, bucket types are converted to affine points at once with [BatchNormalize()](https://github.com/kroma-network/tachyon/blob/522243b7b62b69c552dfa4768d9b7160f9e1694c/tachyon/math/elliptic_curves/short_weierstrass/point_xyzz.h#L113-L159). This reduces the number of required inverse operations to once per sequence.
+
+Relevant commit: [here](https://github.com/kroma-network/tachyon/pull/294/commits/0a9fe9c56c8d7bf9d193c89991003cee3d6a88b9)
+
+### 2.1. Batch Normalize for Permuted Polys
+
+Relevant code optimization: [here](https://github.com/kroma-network/tachyon/pull/294/commits/0a9fe9c56c8d7bf9d193c89991003cee3d6a88b9#diff-a66badbb6b9b05ff643f192101f10cd59fe8edc0d9e63583f088859089c4e80fR150-R165).
+
+Relevant halo2 code: [permuted input poly] and [permuted table poly].
+
+### 2.2. Batch Normalize for Grand Product Polys
+
+Relevant code optimization: [here](https://github.com/kroma-network/tachyon/pull/294/commits/0a9fe9c56c8d7bf9d193c89991003cee3d6a88b9#diff-a66badbb6b9b05ff643f192101f10cd59fe8edc0d9e63583f088859089c4e80fR172-R196).
+
+Relevant halo2 code: [permutation product poly] and [lookup product poly].
+
+## 2.3. Full use of batch inverse
+
+Inverse operations are costly for elliptic curves. We can reduce the number of inverse operations on multiple points by using [batch inverse](https://github.com/kroma-network/tachyon/blob/48891d59d2d751665cb12bc65267d48450c847df/tachyon/math/base/groups.h#L152-L199), which only requires one inverse operation for all the points included.
+
+We fully utilize batch inverse to avoid multiple inverse operations. For example, the original [batch_normalize](https://github.com/zkcrypto/group/blob/0c5b04443b2b24c9a06d50ca890313b641f2a5df/src/lib.rs#L102-L110) inefficiently computed an inverse operation for every point, so we optimized our [BatchNormalize()](https://github.com/kroma-network/tachyon/blob/48891d59d2d751665cb12bc65267d48450c847df/tachyon/math/elliptic_curves/short_weierstrass/point_xyzz.h#L113-L159) to use the batch inverse operation to normalize all the points with a single inverse operation.
```

Comment:
```suggestion
We fully utilize **Batch Inverse** to avoid multiple inverse operations. For example, the original [batch_normalize](https://github.com/zkcrypto/group/blob/0c5b04443b2b24c9a06d50ca890313b641f2a5df/src/lib.rs#L102-L110) inefficiently computed an inverse operation for every point, so we optimized our [BatchNormalize()](https://github.com/kroma-network/tachyon/blob/48891d59d2d751665cb12bc65267d48450c847df/tachyon/math/elliptic_curves/short_weierstrass/point_xyzz.h#L113-L159) to use the batch inverse operation to normalize all the points with a single inverse operation.
```